README
$Id: README,v 1.28 2001/11/04 10:17:59 tujikawa Exp $

Aria - yet another download tool

Copyright (c)2000, 2001 Tatuhiro Tujikawa <tujikawa@rednoah.com>

1. Disclaimer
-------------
This program comes with no warranty.
You must use this program at your own risk.

2. Basic features
-----------------
- Download file via HTTP or FTP
- Tabbed mode
- Aria can avoid annoying CGI download page(Server Template feature)
  that usual download softwares cannot handle properly.
  If you encounter these CGI page, please inform me, then I write
  Server Template for it as long as I can.
- Recursive download feature with verious useful options(HTTP and FTP).
- Drag and Drop support(Mozilla, Galeon, Konqueror, Opera, Netscape).
  In Mozilla, not only url links, plain texts can be dropped.
- Aria can execute command when downloads complete.
- Aria can execute command according to the extention of retrieved file.
- Simple timer feature.
- Graphical progress bar.
- MD5/CRC(both 16bit and 32bit) checking.
- URL numerical expansion feature(also known as numerical download).
- HTTP/FTP proxy support.
- Split download.
- Simple but very useful menu and options.
- Cut and Paste support.
- Cookie support.
- Very stable.

3. Requirements
---------------
C++ compiler(I'm using gcc 2.95.4 and gcc 3.0.2)
libstdc++
GTK version 1.2.6 or higher
Gettext 0.10.35 or higher

4. Compile and install
----------------------
Unpack the archive and enter aria source directory,
	./configure
	make
	make install

Since version 0.10.0, you don't need to run mksetup.
If .aria directory does not exist in your home directory, Aria automatically
creates it and copies necessary files into it.

If you want to initialize these configuration files, run Aria with "--initrc"
option.

I recommend to run Aria with "--init" every time you run the new version of
Aria for the first time.
(Please make back up copies of your setting files(rc.aria, useragent.aria,
server.aria, command.aria) if necessary.

5. If you are using galeon,...
------------------------------
In galeon setting menu, choose "handler->download".
In "Program" frame, choose "use command line"(or something like that),
and enter "aria -g %s" in "Command" entry.
(If you don't install Aria in search path, you may have to enter full
path of aria.)

Now, you can download every thing by Aria from galeon.

6. Configuration
----------------
If you wish to specify the font manually, edit ~/.aria/rc.aria.

The directory ~/.aria/save/ contains several configuration and save file:
Download list is saved to list.aria at the end of every session.

default.aria contains default item option for new list.

http_proxy.aria contains HTTP proxy server information. The syntax of this
file is:
SERVERNAME[(':'|' ')PORT]
If PORT is omitted, 8080 is used. The lines starting '#' are a comment.

useragent.aria containes User-Agent definition. You can add your favorite
User-Agent there. The lines starting '#' are treated as comment lines.

server.aria is a Server Template File. This file defines file retrieving
procedure for specific web servers. Please read resources/server.aria.template
and server.aria.

command.aria is a command list that is executed when each download completes.
Please read resources/command.aria.template.

7. ICON SET
-----------
Icon set is a collection of icons that represents download status(1st column
of download list).

The icon set path is ~/.aria/statusIcon/iconset
Create a directory under it, then place following xpm files in the directory
you created:
  filename     : memo
=====================================
  complete.xpm : download complete
  download.xpm : download in progress
  lock.xpm     : locked
  ready.xpm    : ready
  divide.xpm   : split download
  error.xpm    : error
  stop.xpm     : stop

Note that the size of xpm is 30x25.

You can select your own icon set in application settings menu.
In Display tab, select your favorit icon set, then click "Apply" button.
The entry name of icon set is the name of the directory you created.

8. pixmap for DND basket
------------------------
You can change the pixmap of DND basket.
Put pixmaps under ~/.aria/basket/pixmap

You can select pixmap in application settings menu.
In Basket tab, select pixmap, then click "Apply" button.
The entry name is the filename of pixmap.

Transparent pixmaps are supported.

9. FAQ
------

I. What is URL list?
--------------------
URLs deliminated by '\n' character
for example:
	foobar01.jpg
	foobar02.jpg
	foobar03.jpg

II. Please tell me the format of CRC/MD5 list
----------------------------------
valid format for CRC list is <filename> <CRC> [<filesize>].
<filesize> can be omissible.
e.g.
index.html     7e2a
download.html  cc34

As for MD5 list,
<MD5 checksum> <filename>

e.g.
c6ef4d3efaa05765a2e1823548e5ebba  index.html
f6144ae398cb959b9385cb8c5fbe474d  download.html

III. What is "paste URL list with numerical expansion" means?
-------------------------------------------------
If you copy an phrase "foobar_00-05.jpg", and activate this menu,
You can get following URLs:
	foobar_00.jpg
	foobar_01.jpg
	foobar_02.jpg
	foobar_03.jpg
	foobar_04.jpg
	foobar_05.jpg
You can specify alphabet in variable character like "foobar_aa-zz.jpg"
instead of decimal number. Also, "foobar_[aa-zz].jpg" produces same
results.

IV. What are the defferences between "Open URL list" and "Find Hyplerlink"?
--------------------------------------------------------------------------
If the file you want to read contains just URL list, use "Open URL list".
If the file is html file, use "Find Hyperlink".

10. Contact
----------
E-mail:   tujikawa@rednoah.com
Web Site: http://aria.rednoah.com/

I realy need your feedbacks. If you notice bugs or something,
tell me by email or message board at my web page.


ChangeLog
2002-12-16  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/download.cc (Send_report): now message is instantly appeard
	in console text when executing command(in right-click menu)

2002-12-15  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/main.cc (main): apply the patch from Matej Vela's initialization
	patch.(see Debian Bug report logs - #134394)
	* src/URLcontainer.cc (Find_URL): now space is not recognized as
	delimiter when substituting URL from drag'n dropped text.
	* src/gui_main.cc (Set_speed_scale): fixed the bug that speed limit
	is not set correctly when items are selected in multiple tabs.
	* src/Retrieve.cc (Download_data): fixed miscalculation of average
	speed: long int is not enough to measure it, use unsigned long instead.
	* src/RetrieveHTTP.cc (Download_data_chunked): fixed miscalculation of
	average speed: long int is not enough to measure it, use unsigned long
	instead.
	* src/RetrieveHTTP.cc (Send_Request): user defined referer and URL
	referer are URL-encoded.
	
2002-10-01  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* Now Aria can be compiled with gcc-3.0 and 3.2.
	But with gcc-3.2, it takes long time to write the item and
	configuration files.

2002-09-30  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/gui_download.cc (Download_stop_sub): delete default parameter
	4 and 5. These parameter is defined by prototype declaration.

2002-09-19  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/HTMLparse.cc (find_href): fixed the feature "extensions not to
	download" is not working.

2002-09-07  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/Base64.cc : merged the patch contributed by Emanuele Caratti.
	It fixes base64 encoding bug.
	* po/hu.po: updated hu.po file, by Boldizsar Nagy.
	* po/zh_TW.po: traditional chinese po file added, contributed
	by hardded.
	* po/fr.po: updated by Nicolas Boos.
	* src/RetrieveHTTP.cc, src/RetrieveFTP.cc: now lower size limitation is
	applied even if total size of file is known to be 0.

2002-04-07  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/gui_file.cc: 
	fixed the bug that cause segmentation fault when Aria exits.

2002-04-04  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/RetrieveFTP.cc : 
	fixed the bug in sending "LIST -l" command if no cwd option is enabled
	* src/FTPcontainer.cc :
	fixed the bug that cannot get ip address and port if in PASV mode
	the server returns the address-port string without enclosed with
	'(' ')'.

2002-04-03  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/Command.cc (Is_reserved): 
	$(filename) doesn't mean not full path anymore, it means just the
	file name. I've added 2 new reserved word: $(filepath) and $(url).
	$(filepath) is what the $(filename) meant, that is the full path
	to the downloaded file. $(url) is the url of the item.

2002-04-02  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* po/it.po:
	added Italian translation, thanks to Giovanni Pardini.
	* src/gui_file.cc:
	fixed the timer bug: Assume "Execute command when timer expires"
	option is enabled. Under this assumption,
	if "do not stop download when timer expires"
	option is disabled, when timer expires and there is at least one
	active download, Aria doesn't execute command.
	* src/ItemOption.cc:
	command entry is replaced by text field(though gtk_text is
	known to be buggy). It's handy to write down shell script.
	This idea is from Adam Purkrt.

2002-03-19  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/HTMLparse.cc: 
	added <base href="..."> support
2002-03-17  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/StringHash.cc:
	which was introduced to replace STL map with.
	It made list file reading speed 30% faster.
	
2002-02-13  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/RetrieveFTP.cc (Get_response): 
	rewrite this function
	* src/AppOption.cc:
	fixed timer behavior
	* src/RetrieveFTP.cc:
	added the option "no CWD"
	* src/main.cc:
	added useragent "Mozilla/4.0 (compatible)" at first place in
	useragent list
	
2002-01-20  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/AppOption.cc:
	replaced "Use timer button" with "Use start timer" and "Use stop
	timer." If both "Use start timer" and "Use stop timer" are enabled,
	then stop time is adjusted based on start timer.
	* src/Socket.cc:
	add (char *) cast in order to complie gcc2.96

2001-12-25  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/gui_download_check.cc : 
	fixed the bug that checks all items while the checks of completed items
	are requested.
	* src/ItemStatusDynamic.cc :
	fixed segmentation fault bug: call itemcell->release_Options_Lock()
	while itemcell has been deleted.
	
2001-11-29  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/gui_item_new_item.cc: 
	new url is now parsed by URLcontainer::Find_URL() firstly.
	* src/AppOption.cc:
	Refined "Execute command" timing configulation.
	
2001-11-24  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/RetrieveHTTP.cc (Send_Request), Options.cc, ItemOption.cc: 
	added Accept-Language request field. Adding your mother language to
	this option may be good for ya? Whatever, I set "ja" there ;)
	
	* src/AppOption.cc:
	now users can determin when the command is executed; these are 2
	choices: when all downloads are over, or when timer expires.
	The first case includes the case when the downloads are stopped by
	timer.
	
2001-11-20  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/AppOption.cc : 
	Added option that bypass paste window when the url is dragged
	into basket
	Added option that do not stop download when timer expires.
	* src/RetrieveHTTPS.cc :
	Added https protocol support. OpenSSL required.

2001-11-18  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/RetrieveHTTP.cc: 
	fixed time out bug while getting HTTP header
	* src/Basket.cc:
	added focus-out call back function

2001-11-14  Tatsuhiro Tsujikawa  <tujikawa@rednoah.com>

	* src/RetrieveHTTP.cc: 
	fixed the bug that ignores "save as" field when redirection
	occurs.
	* src/RetrieveHTTP.cc:

src/md5.c
/* md5.c - MD5 Message-Digest Algorithm
 *	Copyright (C) 1995, 1996, 1998, 1999 Free Software Foundation, Inc.
 *
 * according to the definition of MD5 in RFC 1321 from April 1992.
 * NOTE: This is *not* the same file as the one from glibc.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */
/* Written by Ulrich Drepper <drepper@gnu.ai.mit.edu>, 1995.  */
/* heavily modified for GnuPG by <werner.koch@guug.de> */

/* modified for Aria by <tujikawa@rednoah.com> */

/* Test values:
 * ""                  D4 1D 8C D9 8F 00 B2 04  E9 80 09 98 EC F8 42 7E
 * "a"                 0C C1 75 B9 C0 F1 B6 A8  31 C3 99 E2 69 77 26 61
 * "abc                90 01 50 98 3C D2 4F B0  D6 96 3F 7D 28 E1 7F 72
 * "message digest"    F9 6B 69 7D 7C B7 93 8D  52 5A 2F 31 AA F1 61 D0
 * "
 */

//#include <config.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
/*
#include "util.h"
#include "memory.h"
#include "dynload.h"

#include "bithelp.h"
*/
#include "md5.h"

void
md5_init( MD5_CONTEXT *ctx )
{
    ctx->A = 0x67452301;
    ctx->B = 0xefcdab89;
    ctx->C = 0x98badcfe;
    ctx->D = 0x10325476;

    ctx->nblocks = 0;
    ctx->count = 0;
}

/* These are the four functions used in the four steps of the MD5 algorithm
   and defined in the RFC 1321.  The first function is a little bit optimized
   (as found in Colin Plumbs public domain implementation).  */
/* #define FF(b, c, d) ((b & c) | (~b & d)) */
#define FF(b, c, d) (d ^ (b & (c ^ d)))
#define FG(b, c, d) FF (d, b, c)
#define FH(b, c, d) (b ^ c ^ d)
#define FI(b, c, d) (c ^ (b | ~d))


/****************
 * transform n*64 bytes
 */
static void
/*transform( MD5_CONTEXT *ctx, const void *buffer, size_t len )*/
transform( MD5_CONTEXT *ctx, byte *data )
{
    u32 correct_words[16];
    u32 A = ctx->A;
    u32 B = ctx->B;
    u32 C = ctx->C;
    u32 D = ctx->D;
    u32 *cwp = correct_words;

  #ifdef BIG_ENDIAN_HOST
    { int i;
      byte *p2, *p1;
      for(i=0, p1=data, p2=(byte*)correct_words; i < 16; i++, p2 += 4 ) {
	p2[3] = *p1++;
	p2[2] = *p1++;
	p2[1] = *p1++;
	p2[0] = *p1++;
      }
    }
  #else
    memcpy( correct_words, data, 64 );
  #endif


#define OP(a, b, c, d, s, T)					    \
  do								    \
    {								    \
      a += FF (b, c, d) + (*cwp++) + T; 	    \
      a = rol(a, s);						    \
      a += b;							    \
    }								    \
  while (0)

    /* Before we start, one word about the strange constants.
       They are defined in RFC 1321 as

       T[i] = (int) (4294967296.0 * fabs (sin (i))), i=1..64
     */

    /* Round 1.  */
    OP (A, B, C, D,  7, 0xd76aa478);
    OP (D, A, B, C, 12, 0xe8c7b756);
    OP (C, D, A, B, 17, 0x242070db);
    OP (B, C, D, A, 22, 0xc1bdceee);
    OP (A, B, C, D,  7, 0xf57c0faf);
    OP (D, A, B, C, 12, 0x4787c62a);
    OP (C, D, A, B, 17, 0xa8304613);
    OP (B, C, D, A, 22, 0xfd469501);
    OP (A, B, C, D,  7, 0x698098d8);
    OP (D, A, B, C, 12, 0x8b44f7af);
    OP (C, D, A, B, 17, 0xffff5bb1);
    OP (B, C, D, A, 22, 0x895cd7be);
    OP (A, B, C, D,  7, 0x6b901122);
    OP (D, A, B, C, 12, 0xfd987193);
    OP (C, D, A, B, 17, 0xa679438e);
    OP (B, C, D, A, 22, 0x49b40821);

#undef OP
#define OP(f, a, b, c, d, k, s, T)  \
    do								      \
      { 							      \
	a += f (b, c, d) + correct_words[k] + T;		      \
	a = rol(a, s);						      \
	a += b; 						      \
      } 							      \
    while (0)

    /* Round 2.  */
    OP (FG, A, B, C, D,  1,  5, 0xf61e2562);
    OP (FG, D, A, B, C,  6,  9, 0xc040b340);
    OP (FG, C, D, A, B, 11, 14, 0x265e5a51);
    OP (FG, B, C, D, A,  0, 20, 0xe9b6c7aa);
    OP (FG, A, B, C, D,  5,  5, 0xd62f105d);
    OP (FG, D, A, B, C, 10,  9, 0x02441453);
    OP (FG, C, D, A, B, 15, 14, 0xd8a1e681);
    OP (FG, B, C, D, A,  4, 20, 0xe7d3fbc8);
    OP (FG, A, B, C, D,  9,  5, 0x21e1cde6);
    OP (FG, D, A, B, C, 14,  9, 0xc33707d6);
    OP (FG, C, D, A, B,  3, 14, 0xf4d50d87);
    OP (FG, B, C, D, A,  8, 20, 0x455a14ed);
    OP (FG, A, B, C, D, 13,  5, 0xa9e3e905);
    OP (FG, D, A, B, C,  2,  9, 0xfcefa3f8);
    OP (FG, C, D, A, B,  7, 14, 0x676f02d9);
    OP (FG, B, C, D, A, 12, 20, 0x8d2a4c8a);

    /* Round 3.  */
    OP (FH, A, B, C, D,  5,  4, 0xfffa3942);
    OP (FH, D, A, B, C,  8, 11, 0x8771f681);
    OP (FH, C, D, A, B, 11, 16, 0x6d9d6122);
    OP (FH, B, C, D, A, 14, 23, 0xfde5380c);
    OP (FH, A, B, C, D,  1,  4, 0xa4beea44);
    OP (FH, D, A, B, C,  4, 11, 0x4bdecfa9);
    OP (FH, C, D, A, B,  7, 16, 0xf6bb4b60);
    OP (FH, B, C, D, A, 10, 23, 0xbebfbc70);
    OP (FH, A, B, C, D, 13,  4, 0x289b7ec6);
    OP (FH, D, A, B, C,  0, 11, 0xeaa127fa);
    OP (FH, C, D, A, B,  3, 16, 0xd4ef3085);
    OP (FH, B, C, D, A,  6, 23, 0x04881d05);
    OP (FH, A, B, C, D,  9,  4, 0xd9d4d039);
    OP (FH, D, A, B, C, 12, 11, 0xe6db99e5);
    OP (FH, C, D, A, B, 15, 16, 0x1fa27cf8);
    OP (FH, B, C, D, A,  2, 23, 0xc4ac5665);

    /* Round 4.  */
    OP (FI, A, B, C, D,  0,  6, 0xf4292244);
    OP (FI, D, A, B, C,  7, 10, 0x432aff97);
    OP (FI, C, D, A, B, 14, 15, 0xab9423a7);
    OP (FI, B, C, D, A,  5, 21, 0xfc93a039);
    OP (FI, A, B, C, D, 12,  6, 0x655b59c3);
    OP (FI, D, A, B, C,  3, 10, 0x8f0ccc92);
    OP (FI, C, D, A, B, 10, 15, 0xffeff47d);
    OP (FI, B, C, D, A,  1, 21, 0x85845dd1);
    OP (FI, A, B, C, D,  8,  6, 0x6fa87e4f);
    OP (FI, D, A, B, C, 15, 10, 0xfe2ce6e0);
    OP (FI, C, D, A, B,  6, 15, 0xa3014314);
    OP (FI, B, C, D, A, 13, 21, 0x4e0811a1);
    OP (FI, A, B, C, D,  4,  6, 0xf7537e82);
    OP (FI, D, A, B, C, 11, 10, 0xbd3af235);
    OP (FI, C, D, A, B,  2, 15, 0x2ad7d2bb);
    OP (FI, B, C, D, A,  9, 21, 0xeb86d391);

    /* Put checksum in context given as argument.  */
    ctx->A += A;
    ctx->B += B;
    ctx->C += C;
    ctx->D += D;
}



/* The routine updates the message-digest context to
 * account for the presence of each of the characters inBuf[0..inLen-1]
 * in the message whose digest is being computed.
 */
void
md5_write( MD5_CONTEXT *hd, byte *inbuf, size_t inlen)
{
    if( hd->count == 64 ) { /* flush the buffer */
	transform( hd, hd->buf );
	hd->count = 0;
	hd->nblocks++;
    }
    if( !inbuf )
	return;
    if( hd->count ) {
	for( ; inlen && hd->count < 64; inlen-- )
	    hd->buf[hd->count++] = *inbuf++;
	md5_write( hd, NULL, 0 );
	if( !inlen )
	    return;
    }

    while( inlen >= 64 ) {
	transform( hd, inbuf );
	hd->count = 0;
	hd->nblocks++;
	inlen -= 64;
	inbuf += 64;
    }
    for( ; inlen && hd->count < 64; inlen-- )
	hd->buf[hd->count++] = *inbuf++;

}



/* The routine final terminates the message-digest computation and
 * ends with the desired message digest in mdContext->digest[0...15].
 * The handle is prepared for a new MD5 cycle.
 * Returns 16 bytes representing the digest.
 */

void
md5_final( MD5_CONTEXT *hd )
{
    u32 t, msb, lsb;
    byte *p;

    md5_write(hd, NULL, 0); /* flush */;

    msb = 0;
    t = hd->nblocks;
    if( (lsb = t << 6) < t ) /* multiply by 64 to make a byte count */
	msb++;
    msb += t >> 26;
    t = lsb;
    if( (lsb = t + hd->count) < t ) /* add the count */
	msb++;
    t = lsb;
    if( (lsb = t << 3) < t ) /* multiply by 8 to make a bit count */
	msb++;
    msb += t >> 29;

    if( hd->count < 56 ) { /* enough room */
	hd->buf[hd->count++] = 0x80; /* pad */
	while( hd->count < 56 )
	    hd->buf[hd->count++] = 0;  /* pad */
    }
    else { /* need one extra block */
	hd->buf[hd->count++] = 0x80; /* pad character */
	while( hd->count < 64 )
	    hd->buf[hd->count++] = 0;
	md5_write(hd, NULL, 0);  /* flush */;
	memset(hd->buf, 0, 56 ); /* fill next block with zeroes */
    }
    /* append the 64 bit count */
    hd->buf[56] = lsb	   ;
    hd->buf[57] = lsb >>  8;
    hd->buf[58] = lsb >> 16;
    hd->buf[59] = lsb >> 24;
    hd->buf[60] = msb	   ;
    hd->buf[61] = msb >>  8;
    hd->buf[62] = msb >> 16;
    hd->buf[63] = msb >> 24;
    transform( hd, hd->buf );

    p = hd->buf;
  #ifdef BIG_ENDIAN_HOST
    #define X(a) do { *p++ = hd->##a	  ; *p++ = hd->##a >> 8;      \
		      *p++ = hd->##a >> 16; *p++ = hd->##a >> 24; } while(0)
  #else /* little endian */
    /*#define X(a) do { *(u32*)p = hd->##a ; p += 4; } while(0)*/
    /* Unixware's cpp doesn't like the above construct so we do it his way:
     * (reported by Allan Clark) */
    #define X(a) do { *(u32*)p = (*hd).a ; p += 4; } while(0)
  #endif
    X(A);
    X(B);
    X(C);
    X(D);
  #undef X

}

static byte *
md5_read( MD5_CONTEXT *hd )
{
    return hd->buf;
}

/****************
 * Return some information about the algorithm.  We need algo here to
 * distinguish different flavors of the algorithm.
 * Returns: A pointer to string describing the algorithm or NULL if
 *	    the ALGO is invalid.
 */

#ifndef IS_MODULE
static
#endif
const char * const gnupgext_version = "MD5 ($Revision: 1.1 $)";
/*
static struct {
    int class;
    int version;
    int  value;
    void (*func)(void);
} func_table[] = {
    { 10, 1, 0, (void(*)(void))md5_get_info },
    { 11, 1, 1 },
};
*/
/*
#ifndef IS_MODULE
static
#endif
void *
gnupgext_enum_func( int what, int *sequence, int *class, int *vers )
{
    void *ret;
    int i = *sequence;

    do {
	if( i >= DIM(func_table) || i < 0 )
	    return NULL;
	*class = func_table[i].class;
	*vers  = func_table[i].version;
	switch( *class ) {
	  case 11: case 21: case 31: ret = &func_table[i].value; break;
	  default:		     ret = func_table[i].func; break;
	}
	i++;
    } while( what && what != *class );

    *sequence = i;
    return ret;
}
*/
/*
#ifndef IS_MODULE
void
md5_constructor(void)
{
    register_internal_cipher_extension( gnupgext_version, gnupgext_enum_func );
}
#endif
*/

/* end of file */

